### PR TITLE
[Fireperf][AASA] send `_experiment_app_start_ttid` trace, controlled by RC flag

### PR DESCRIPTION
**Pull request by leotianlizhan**
b/245947422
b/243540978

During startup, when we want to decide to record this trace or not, we only check disk cache for RC flag because RC is not initialized yet. 

This RC flag will be saved to disk ONLY upon successful RC `fetchAndActivate()`. There's a problem with all other flags in that they are being saved to disk every time they are queried.

### CODE REVIEW

File: firebase-perf/dev-app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -38,6 +38,9 @@
         <meta-data
             android:name="fragment_sampling_percentage"
             android:value="100.0" />
+        <meta-data
+            android:name="experiment_as_ttid"
```
Comment (raymondlam) [inline firebase-perf/dev-app/src/main/AndroidManifest.xml:42]: I'd prefer experiment_app_start_ttid so that it is more obvious that it is for app start.

File: firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigurationConstants.java [hunk 2]
```diff
@@ -661,4 +661,39 @@ protected String getMetadataFlag() {
...
+    }
+
+    @Override
+    protected Boolean getDefault() {
+      return false;
+    }
+
+    @Override
+    protected String getRemoteConfigFlag() {
+      return "fpr_experiment_as_ttid";
```
Comment (raymondlam) [inline firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigurationConstants.java:686]: Similarly, `as` should be `app_start` here.

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java [hunk 3]
```diff
@@ -227,11 +272,24 @@ public synchronized void onActivityResumed(Activity activity) {
 
     // Log the app start trace in a non-main thread.
     executorService.execute(this::logAppStartTrace);
+  }
 
-    if (isRegisteredForLifecycleCallbacks) {
-      // After AppStart trace is logged, we can unregister this callback.
-      unregisterActivityLifecycleCallbacks();
```
Comment (raymondlam) [inline firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:233]: Are we still unregistering somewhere if the the experiment is disabled and recordFirstDrawDone never gets called?

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java [hunk 4]
```diff
@@ -227,11 +272,24 @@ public synchronized void onActivityResumed(Activity activity) {
...
+  private void logColdStart(Timer start, Timer end, PerfSession session) {
+    TraceMetric.Builder metric =
+        TraceMetric.newBuilder()
+            .setName("_experiment_as_ttid")
+            .setClientStartTimeUs(start.getMicros())
+            .setDurationUs(start.getDurationMicros(end));
+
+    TraceMetric.Builder subtrace =
+        TraceMetric.newBuilder()
+            .setName("_experiment_classLoadTime")
```
Comment (raymondlam) [inline firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:286]: What's the purpose of this sub trace?

File: firebase-perf/src/test/java/com/google/firebase/perf/metrics/AppStartTraceTest.java [hunk 5]
```diff
@@ -261,4 +279,38 @@ public void testFirebasePerfProviderOnAttachInfo_initializesGaugeCollection() {
     Assert.assertEquals(oldSessionId, SessionManager.getInstance().perfSession().sessionId());
     verify(mockPerfSession, times(2)).isGaugeAndEventCollectionEnabled();
   }
+
+  @Test
+  @Config(sdk = 26)
+  public void timeToInitialDisplay_isLogged() {
```
Comment (raymondlam) [inline firebase-perf/src/test/java/com/google/firebase/perf/metrics/AppStartTraceTest.java:285]: Can you add some comments to clarify what is going on in this test? There is a lot of setup and some verify statements mid way through the test making it hard to follow.

File: firebase-perf/dev-app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (leotianlizhan) [reply firebase-perf/dev-app/src/main/AndroidManifest.xml]: done

File: firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigurationConstants.java (hunk 2 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigurationConstants.java]: done

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 3 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:233]: Great catch! Added it back.

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 4 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:292]: This captures the old-start, which is class-load of `FirebasePerfProvider.class`. I just want to have more data to study the difference that the new start would have.

File: firebase-perf/src/test/java/com/google/firebase/perf/metrics/AppStartTraceTest.java (hunk 5 above)
Comment (leotianlizhan) [reply firebase-perf/src/test/java/com/google/firebase/perf/metrics/AppStartTraceTest.java:285]: Added comments! The setup is quite big, only the last few lines of this test is actually verifying.

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java [hunk 6]
```diff
@@ -228,12 +274,30 @@ public synchronized void onActivityResumed(Activity activity) {
...
+    if (!isExperimentTTIDEnabled && isRegisteredForLifecycleCallbacks) {
       // After AppStart trace is logged, we can unregister this callback.
       unregisterActivityLifecycleCallbacks();
     }
   }
 
+  private void logColdStart(Timer start, Timer end, PerfSession session) {
+    TraceMetric.Builder metric =
+        TraceMetric.newBuilder()
+            .setName("_experiment_app_start_ttid")
```
Comment (jeremyjiang-dev) [inline firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:286]: nit: define the name as a constant Constants.TraceNames.

File: firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java [hunk 7]
```diff
@@ -767,6 +768,32 @@ public float getFragmentSamplingRate() {
...
+    // 4. Otherwise, return default value.
+    ExperimentTTID config = ExperimentTTID.getInstance();
+
+    // 1. Reads value in Android Manifest (it is set by developers during build time).
+    Optional<Boolean> metadataValue = getMetadataBoolean(config);
+    if (metadataValue.isAvailable()) {
+      return metadataValue.get();
+    }
+
+    // 2. Cannot read value from RC because it's not initialized yet.
```
Comment (jeremyjiang-dev) [inline firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java:786]: Just to make sure my understanding is correct: for the very first app start of an app, we will read the default value.

File: firebase-perf/dev-app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -38,6 +38,9 @@
         <meta-data
             android:name="fragment_sampling_percentage"
             android:value="100.0" />
+        <meta-data
```
Comment (jeremyjiang-dev) [inline firebase-perf/dev-app/src/main/AndroidManifest.xml:41]: Can you also update the e2e test app so that it will generate this trace periodically? but I think our e2e test app does not generate app start trace. If so, it is out of the scope of this PR.

**Review by jeremyjiang-dev — APPROVED** [review]
Looks good to me overall

File: firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java (hunk 7 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java:786]: Yes, for the very very first without any prior RC fetch, it will use default value.

File: firebase-perf/dev-app/src/main/AndroidManifest.xml (hunk 8 above)
Comment (leotianlizhan) [reply firebase-perf/dev-app/src/main/AndroidManifest.xml:41]: Checked just now and you are right, our E2E test app doesn't generate `_app_start` at all.

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 6 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:286]: This code is for dark-launch, it will be removed next release, so I didn't add a new official constant to Constants.java. I will definitely do it for the GA launch.

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java [hunk 9]
```diff
@@ -228,12 +274,30 @@ public synchronized void onActivityResumed(Activity activity) {
     // Log the app start trace in a non-main thread.
     executorService.execute(this::logAppStartTrace);
 
-    if (isRegisteredForLifecycleCallbacks) {
+    if (!isExperimentTTIDEnabled && isRegisteredForLifecycleCallbacks) {
```
Comment (visumickey) [inline firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:277]: What happens if the flag dynamically changes from No to YES (or) vice versa? Would that cause a situation where we try to unregister for callbacks when we have not registered or the other way about?

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java [hunk 10]
```diff
@@ -228,12 +274,30 @@ public synchronized void onActivityResumed(Activity activity) {
...
+
+    TraceMetric.Builder subtrace =
+        TraceMetric.newBuilder()
+            .setName("_experiment_classLoadTime")
+            .setClientStartTimeUs(FirebasePerfProvider.getAppStartTime().getMicros())
+            .setDurationUs(FirebasePerfProvider.getAppStartTime().getDurationMicros(end));
+
+    metric.addSubtraces(subtrace).addPerfSessions(this.startSession.build());
+
+    transportManager.log(metric.build(), ApplicationProcessState.FOREGROUND_BACKGROUND);
```
Comment (visumickey) [inline firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:298]: Why FOREGROUND_BACKGROUND?

File: firebase-perf/dev-app/src/main/AndroidManifest.xml (hunk 8 above)
Comment (jeremyjiang-dev) [reply firebase-perf/dev-app/src/main/AndroidManifest.xml:41]: IIRC the test skips the onResume() callback

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 4 above)
Comment (raymondlam) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:292]: sg

**Review by raymondlam — APPROVED** [review]

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 10 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:298]: `_app_start` is `FOREGROUND_BACKGROUND` so I am just following the same behavior.

File: firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java [hunk 11]
```diff
@@ -767,6 +768,32 @@ public float getFragmentSamplingRate() {
     return config.getDefault();
   }
 
+  /** Returns if _experiment_app_start_ttid should be captured. */
+  public boolean getIsExperimentTTIDEnabled() {
+    // Order of precedence is:
+    // 1. If the value exists in Android Manifest, return this value.
+    // 2. Cannot read value from RC because it's not initalized yet
```
[the remark this answers is not in the document]
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java:775]: This is a comment in the same style as other flags if you look into the same file. Other flags has the order laid out like this, I just changed the 2nd step to "Cannot read from RC".

File: firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java [hunk 12]
```diff
@@ -767,6 +768,32 @@ public float getFragmentSamplingRate() {
     return config.getDefault();
   }
 
+  /** Returns if _experiment_app_start_ttid should be captured. */
+  public boolean getIsExperimentTTIDEnabled() {
+    // Order of precedence is:
```
[the remark this answers is not in the document]
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/config/ConfigResolver.java:773]: This is just Manifest > Device Cache > SDK Default

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 9 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:277]: Good question! Since this flag is only read once in the beginning of `AppStartTrace#onActivityResumed`, it's not a problem.
> Scenario 1: experiment disabled, an activity resumed, experiment enabled, another activity resumed.

After the first resume, the unregistering would be called, so the 2nd activity resume would not do anything. TTID isn't recorded for this app start, it will only start working next app start.

> Scenario 2: experiment enabled, an activity resumed, experiment enabled, another activity resumed.

After the first resume, the `AppStartTrace.onResume` variable would be set, so even if the 2nd activity resume triggers, nothing happens because this method will return immediately without sending anything else.

File: firebase-perf/src/test/java/com/google/firebase/perf/config/RemoteConfigManagerTest.java [hunk 13]
```diff
@@ -230,6 +240,29 @@ public void getString_validFrcValue_updatesWithNewValue() {
     assertThat(testRemoteConfigManager.getString("some_key3").isAvailable()).isFalse();
   }
 
+  @Test
+  public void syncConfigValues_savesNewlyFetchedValueToDeviceCache() {
```
[the remark this answers is not in the document]
Comment (leotianlizhan) [reply firebase-perf/src/test/java/com/google/firebase/perf/config/RemoteConfigManagerTest.java:244]: It is already done in this test. `syncConfigValues` simulates fetch success (it is only called after fetch success), so in this test we call `syncConfigValues` twice and verifies that the disk cache value has changed.

File: firebase-perf/src/main/java/com/google/firebase/perf/config/RemoteConfigManager.java [hunk 14]
```diff
@@ -342,6 +341,19 @@ protected void syncConfigValues(Map<String, FirebaseRemoteConfigValue> newlyFetc
         allRcConfigMap.remove(existingKey);
       }
     }
+
+    // Save to device cache upon successful RC fetchAndActivate
```
[the remark this answers is not in the document]
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/config/RemoteConfigManager.java]: Done

File: firebase-perf/src/main/java/com/google/firebase/perf/config/RemoteConfigManager.java [hunk 15]
```diff
@@ -315,7 +314,7 @@ private void triggerRemoteConfigFetchIfNecessary() {
       return;
     }
     if (allRcConfigMap.isEmpty()) { // Initial fetch.
-      syncConfigValues(firebaseRemoteConfig.getAll());
+      allRcConfigMap.putAll(firebaseRemoteConfig.getAll());
```
[the remark this answers is not in the document]
Comment (visumickey) [reply firebase-perf/src/main/java/com/google/firebase/perf/config/RemoteConfigManager.java:317]: Sounds good. That makes sense.

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 9 above)
Comment (visumickey) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:277]: Sounds good.

For the first scenario - I assume that experiment is disabled?

In the second scenario - I assume that would mean experiment is enabled?

**Review by visumickey — APPROVED** [review]

File: firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java (hunk 9 above)
Comment (leotianlizhan) [reply firebase-perf/src/main/java/com/google/firebase/perf/metrics/AppStartTrace.java:277]: Yes, these scenarios show what happens if the app starts with 1 state, then changes after app start.

**Comment by leotianlizhan** [comment]
`Health Metrics / Coverage (push)` fails due to firestorage test failures, so we can merge.
